syntax = "proto3";

package vecsearchgrpc;
//option go_package = "./recommender/par2vec";

service VectorSearchGrpc {
    rpc createIndex(CreateIndexRequest)     returns (DefaultReply);
    rpc deleteIndex(DefaultRequest)         returns (DefaultReply);

    rpc loadIndex(DefaultRequest)           returns (DefaultReply);
    rpc unloadIndex(DefaultRequest)         returns (DefaultReply);

    rpc searchNeighbors(SearchRequest)     returns (SearchReply) {};
}

message CreateIndexRequest {
    string   indexName  = 1;
    uint64   numVectors = 2;
    uint32   dim        = 3;
    repeated float vecData = 4;
}

message SearchRequest {
    string  indexName = 1;
    uint32  numQueryVectors = 2;
    uint32  dim = 3;
    repeated float vecData = 4;
    uint32  numNeighbors = 5;  
}

message SearchReply {
    string status = 1;
    string message = 2;

    uint32 numQueryVectors = 3;
    uint32 numNeighbors = 4;
    repeated int64 I = 5;
    repeated float  D = 6;
}

message DefaultRequest {
    string  indexName  = 1;
}

message DefaultReply {
    string status = 1;
    string message = 2;
}